Course Updates date validation

efischer19 opened the pull request:
TNL-4115. Previously, course updates (which are intended to be posted with
dates, for sorting in the LMS) could be authored in studio with a valid
date, nothing, or a random string as the "date" for the update. As there
is no validation for this in studio, everything succeeded with no warning.

However, the LMS has problems parsing some of these values, and barfs when
loaded by learners.

The fix does two big things:
- gracefully handles invalid dates in LMS. These updates are now treated as
having a date of today, for sorting purposes.
- turns on validation in studio. Now, it is not only impossible to enter
invalid dates in studio, but notifications will draw the course author's
eye if any invalid updates were previously saved.

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

efischer19 commented:
The big thing remaining is validation error handling - `course_info_update` neither displays nor clears these errors nicely. Otherwise, this is looking half-decent.

efischer19 commented:
Prior behavior screenshots. Before this fix, a date picker exists when adding or editing a course update, but you can also just type anything you want into that box. This looks fine to the course author in studio, but has the effect of breaking all course updates in the LMS.

Broken LMS:
<img width="1188" alt="screen shot 2016-04-05 at 12 16 38 pm" src="https://cloud.githubusercontent.com/assets/7373924/14289827/844f3440-fb2a-11e5-850b-350211ded1fa.png">

Doesn't-appear-broken Studio:
<img width="1315" alt="screen shot 2016-04-05 at 12 17 05 pm" src="https://cloud.githubusercontent.com/assets/7373924/14289845/93f8fa48-fb2a-11e5-8f76-f7f6a56e93e4.png">
<img width="1339" alt="screen shot 2016-04-05 at 12 17 17 pm" src="https://cloud.githubusercontent.com/assets/7373924/14289846/984317e6-fb2a-11e5-9903-ad9900cc6ddc.png">
<img width="1359" alt="screen shot 2016-04-05 at 12 17 28 pm" src="https://cloud.githubusercontent.com/assets/7373924/14289854/9bc273c6-fb2a-11e5-8384-2fe32f57ee53.png">

efischer19 commented:
Post-fix screenshots.

After the fix, the LMS will not explode when it encounters an invalid date. The issue was in code that was trying to sort updates by date, so when errors are encountered, the "sort by date" is assumed to be today by default. In addition, Studio will now show warnings about invalid dates to the course author.

Fixed LMS:
<img width="1186" alt="screen shot 2016-04-05 at 12 34 48 pm" src="https://cloud.githubusercontent.com/assets/7373924/14289956/293afc14-fb2b-11e5-99b7-d0efc581faa5.png">
<img width="929" alt="screen shot 2016-04-05 at 12 35 15 pm" src="https://cloud.githubusercontent.com/assets/7373924/14289959/2d626d36-fb2b-11e5-8993-5c558bc53f82.png">
<img width="1231" alt="screen shot 2016-04-05 at 12 35 25 pm" src="https://cloud.githubusercontent.com/assets/7373924/14289964/31c613c8-fb2b-11e5-88b1-977ba5d3a638.png">
This behavior is live at https://efischer19.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/about.

Fixed studio:
<img width="1303" alt="screen shot 2016-04-05 at 12 38 53 pm" src="https://cloud.githubusercontent.com/assets/7373924/14290027/73d301d6-fb2b-11e5-81fe-713ba17d9d74.png">
<img width="948" alt="screen shot 2016-04-05 at 12 39 08 pm" src="https://cloud.githubusercontent.com/assets/7373924/14290031/78c5d0a6-fb2b-11e5-80bc-1ddd09117212.png">
<img width="919" alt="screen shot 2016-04-05 at 12 39 20 pm" src="https://cloud.githubusercontent.com/assets/7373924/14290041/7fb16952-fb2b-11e5-9e36-fa22facced9f.png">
<img width="920" alt="screen shot 2016-04-05 at 12 40 26 pm" src="https://cloud.githubusercontent.com/assets/7373924/14290069/9a6d6426-fb2b-11e5-8962-8492fc1d90dc.png">

efischer19 commented:
@catong - This is the PR I was mentioning earlier in standup. I'll swing by sometime this afternoon to discuss any potential doc changes, if they're needed.

@chris-mike I was not able to use the pattern library for this (per @andy-armstrong), so I copied the errors used by course settings dates. I think the modal dialog looks alright, but the main page could use some attention. Any suggestions?

@cahrens @dianakhuang Can you give this PR a TNL review? Notes are above, sandbox https://efischer19.sandbox.edx.org/ is live with my changes.

Note: I realized while taking screenshots that invalid-but-not-empty values will pre-populate the date dialog with `Nan/Nan/Nan`, I'm already working on fixing that.

dianakhuang commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -448,6 +448,16 @@ def get_html(self):
+    @staticmethod
+    def safe_parse_date(date):
+        """
+        Since this is used soley for ordering purposes, use today's date as a default
```
Nit: spelling of 'soley'

cahrens commented on cms/static/js/models/course_update.js:
```diff
@@ -6,6 +6,14 @@ define(["backbone", "jquery", "jquery.ui"], function(Backbone, $) {
+        },
+        validate: function(attrs) {
+            var date_exists = (attrs.date !== null && attrs.date !== "");
+            var date_is_valid_string = ($.datepicker.formatDate('MM d, yy', new Date(attrs.date)) === attrs.date);
```
I'm confused about date_is_valid_string and date_is_date. Is date_is_date redundant?

efischer19 replied:
It is, I've removed it in an updated commit

cahrens commented on cms/static/js/utils/date_utils.js:
```diff
@@ -1,15 +1,80 @@
-        var date = $(datepickerInput).datepicker("getDate");
-        var time = $(timepickerInput).timepicker("getTime");
+        var date = null, time = null;
+        if (datepickerInput) {
```
When is datepickerInput ever falsey? If it is legitimate for it to not be present, it seems that you would need to do the same length check that you do for timepickerInput. But if it really never happens, let's just remove the check.

cahrens commented on cms/static/js/views/course_info_update.js:
```diff
@@ -27,22 +30,58 @@ define(["js/views/baseview", "codemirror", "js/models/course_update",
+            }
+            var targetModel = this.collection.get(this.$currentPost.attr('name'));
+            var prevValue = targetModel.get(attr);
+            if (prevValue === value) { return; }
```
Nit: a clean code principle is to avoid multiple points of return. I think an exception to this is error checking at the very top of the method. But in this case, I think the code would be cleaner to do this:

````
if (prevValue !=== value) {
    targetModel.set(attr, value);
    this.validateModel(targetModel);
}
````

cahrens commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -448,6 +448,16 @@ def get_html(self):
             return self.system.render_template("{0}/course_updates.html".format(self.TEMPLATE_DIR), context)
 
+    @staticmethod
+    def safe_parse_date(date):
```
Can you please add a unit test covering this behavior?

cahrens commented:
@efischer19 I'm done with my first pass of this PR. There is no test coverage, so please see what is reasonable in regards to that.

efischer19 commented on common/lib/xmodule/xmodule/tests/test_html_module.py:
```diff
@@ -144,3 +144,24 @@ def test_index_dictionary_html_module_with_script_and_style_tags(self):
             "content": {"html_content": " This has HTML comment in it. HTML end. ", "display_name": "Text"},
             "content_type": "Text"
         })
+
```
I verified the following test fails when run against unfixed master code:

'''
edxapp@precise64:~/edx-platform$ nosetests common/lib/xmodule/xmodule/tests/test_html_module.py:CourseInfoModuleTestCase --id-file /edx/app/edxapp/edx-platform/.testids/lms/noseids --xunit-file /edx/app/edxapp/edx-platform/reports/lms/nosetests.xml --verbosity=1 --with-xunit --xunit-file=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml
F
-----------------------------------------------------------------------------
1) FAIL: testUpdatesRender (xmodule.tests.test_html_module.CourseInfoModuleTestCase)

   Traceback (most recent call last):
    common/lib/xmodule/xmodule/tests/test_html_module.py line 167 in testUpdatesRender
      self.fail("CourseInfoModule could not parse an invalid date!")
   AssertionError: CourseInfoModule could not parse an invalid date!
'''

efischer19 commented on cms/static/coffee/spec/views/course_info_spec.coffee:
```diff
@@ -41,6 +41,7 @@ define ["js/views/course_info_handout", "js/views/course_info_update", "js/model
 
                 @event = {
                     preventDefault : () -> 'no op'
+                    currentTarget : @courseInfoEdit.$el.find('.editing')
```
This can be removed

efischer19 commented on cms/static/coffee/spec/views/course_info_spec.coffee:
```diff
@@ -58,7 +59,7 @@ define ["js/views/course_info_handout", "js/views/course_info_update", "js/model
                     spyOn(model, "save").andCallThrough()
 
-                    cancelEditingUpdate(@courseInfoEdit, @courseInfoEdit.$modalCover, useCancelButton)
+                    @cancelEditingUpdate(@courseInfoEdit, @courseInfoEdit.$modalCover, useCancelButton)
```
Same for all these changes

efischer19 commented:
@cahrens @dianakhuang This is ready for re-review. I'm updating the sandbox right now.

The latest 3 commits contain all recent changes if you want to review them one at a time.

The final commit regarding styling is still in-flight pending input from @chris-mike

cahrens commented on cms/static/js/utils/date_utils.js:
```diff
@@ -61,7 +61,7 @@ function($, date, TriggerChangeEventOnEnter) {
         // UTC timezone, NOT the timezone of the user's browser.
         var date = null, time = null;
-        if (datepickerInput) {
+        if (datepickerInput.length > 0) {
```
Get rid of this check-- you have to at least have a datepicker for this utility method to be useful.

cahrens commented on cms/static/coffee/spec/views/course_info_spec.coffee:
```diff
@@ -153,6 +153,19 @@ define ["js/views/course_info_handout", "js/views/course_info_update", "js/model
+                @courseInfoEdit.onNew(@event)
+                @courseInfoEdit.$el.find('input.date').val("Marchtober 40, 2048")
+                @courseInfoEdit.$el.find('input.date').change()
+                expect(@courseInfoEdit.$el.find('.save-button').hasClass("is-disabled")).toEqual(true)
```
Extend the test in these ways:

1) Verify that Save button is enabled on initial load (assuming it default to current date).
2) After verifying that the button disables after entering an invalid date, change the date to something valid and make sure it enables the Save button.

cahrens commented on cms/static/js/views/course_info_update.js:
```diff
@@ -75,15 +124,15 @@ define(["js/views/baseview", "codemirror", "js/models/course_update",
 
-            $('.date').datepicker('destroy');
-            $('.date').datepicker({ 'dateFormat': 'MM d, yy' });
+            DateUtils.setupDatePicker("date", this, 0);
```
Nit: add a comment about what "0" is here.

cahrens commented:
👍 On the code and functionality. However, I'm not at all crazy about the styling. Please sync up with @chris-mike.

![image](https://cloud.githubusercontent.com/assets/484484/14358604/adaad082-fcbb-11e5-97a7-33a45f14206b.png)

dianakhuang commented:
:+1: On the code changes.

efischer19 commented:
@cahrens @dianakhuang I've gotten the styles fixed up (with help from @chris-mike), and will merge once the tests have passed.

Only changes since your reviews are styling and the error string value.

<img width="965" alt="screen shot 2016-04-08 at 2 03 23 pm" src="https://cloud.githubusercontent.com/assets/7373924/14393174/6c765f68-fd93-11e5-8468-8cb69d5df6a1.png">
<img width="907" alt="screen shot 2016-04-08 at 2 03 16 pm" src="https://cloud.githubusercontent.com/assets/7373924/14393180/721be118-fd93-11e5-9993-5b4af408fd9c.png">
